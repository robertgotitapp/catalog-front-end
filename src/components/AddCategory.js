import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { Redirect } from 'react-router-dom';
import { addCategory } from '../actions/categories';

export class AddCategory extends Component {
    state = {
      name: '',
      description: '',
      toHome: false,
      alerts: '',
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    validateInput = () => {
      const {
        name, description,
      } = this.state;
      const errors = {};
      if (name.length <= 5) {
        errors.name = 'Name must be longer than 5 characters.';
      }
      if (description.length > 200) {
        errors.description = 'Description must be within 200 characters.';
      }
      if (Object.keys(errors).length !== 0) {
        this.setState({
          alerts: errors,
        });
        return false;
      }
      return true;
    }

    handleSubmit = (e) => {
      e.preventDefault();
      const { name, description } = this.state;
      if (this.validateInput()) {
        this.props.addCategory({ name, description })
          .then((res) => {
            if (res.statusCode) {
              this.setState(prevState => ({ ...prevState, toHome: true }));
            } else {
              // If add category request is failed, reset all the form input
              // and display all the error messages as alerts
              res.errorPromise
                .then((error) => {
                  this.setState({
                    name: '',
                    description: '',
                    toHome: false,
                    alerts: error.message,
                  });
                });
            }
          });
      }
    }

    render() {
      const { name, description } = this.state;
      if (this.state.toHome) {
        return <Redirect to='/' />;
      }

      return (
        <div>
          <h2> Add Category </h2>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Category Name</Form.Label>
              <Form.Control type="text" name="name" onChange={this.handleChange} value={name} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Category Description</Form.Label>
              <Form.Control type="text" name="description" onChange={this.handleChange} value={description} />
            </Form.Group>
            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
          {
            // Loop through alerts dictionary and display an alert for each
            // of one item in dictionary
            this.state.alerts
            && Object.keys(this.state.alerts).map((key, id) => (
              <Alert key={id} variant='danger'>
                {key}
                :
                {this.state.alerts[key]}
              </Alert>
            ))
          }
        </div>
      );
    }
}

function mapStateToProps({ usersReducer }) {
  return {
    users: usersReducer,
  };
}

const mapDispatchToProps = {
  addCategory,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddCategory);
