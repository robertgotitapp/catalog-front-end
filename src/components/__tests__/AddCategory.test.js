import React from 'react';
import { shallow } from 'enzyme';
import Form from 'react-bootstrap/Form';
import { AddCategory } from '../AddCategory';

describe('components/AddCategory', () => {
  const props = {};
  const container = shallow(<AddCategory {...props} />);

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should contain 1 FormControl for name', () => {
    expect(container.find('[name="name"]')).toHaveLength(1);
  });

  it('should contain 1 FormControl for description', () => {
    expect(container.find('[name="description"]')).toHaveLength(1);
  });

  it('should trigger state change when FormControl name got input', () => {
    const categoryNameInput = container.find('[name="name"]');
    categoryNameInput.simulate('change', { target: { name: 'name', value: 'category1' } });
    expect(container.state().name).toEqual('category1');
  });

  it('should trigger state change when FormControl description got input', () => {
    const categoryDesscriptionInput = container.find('[name="description"]');
    categoryDesscriptionInput.simulate('change', { target: { name: 'description', value: 'description' } });
    expect(container.state().description).toEqual('description');
  });

  it('should trigger handler when form is submited', () => {
    container.instance().handleSubmit = jest.fn();
    container.instance().forceUpdate();
    const form = container.find(Form);
    form.simulate('submit');
    expect(container.instance().handleSubmit).toHaveBeenCalled();
  });
});
