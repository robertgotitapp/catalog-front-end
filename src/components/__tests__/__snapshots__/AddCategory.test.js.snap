// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddCategory should render correctly 1`] = `
<div>
  <Form
    inline={false}
    onSubmit={[Function]}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Category Name
      </FormLabel>
      <FormControl
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Category Description
      </FormLabel>
      <FormControl
        name="description"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <Button
      active={false}
      disabled={false}
      type="submit"
      variant="primary"
    >
      Submit
    </Button>
  </Form>
</div>
`;
